nilearn/image/image.py

```diff
@@ -502,6 +502,65 @@ def swap_img_hemispheres(img):
 
     # create swapped nifti object
     out_img = nibabel.Nifti1Image(img.get_data()[::-1], img.get_affine(),
-        header=img.get_header())
+                                  header=img.get_header())
 
     return out_img
+
+
+def _index_img(imgs, index):
+    """Helper function for index_img and iter_img
+    """
+    return nibabel.Nifti1Image(imgs.get_data()[..., index],
+                               imgs.get_affine(),
+                               header=imgs.get_header())
+
+
+def index_img(imgs, index):
+    """Indexes into a 4D Niimg-like object in the fourth dimension
+
+    Parameters
+    ----------
+    imgs: 4D Niimg-like object
+          See http://nilearn.github.io/building_blocks/manipulating_mr_images.html#niimg.
+
+    index: Any type compatible with numpy array indexing
+           Used for indexing the 4D data array in the fourth dimension
+
+    Returns
+    -------
+    output: nibabel.Nifti1IMage
+            4D image whose data is a subset of the original image data
+
+    """
+    imgs = check_niimgs(imgs)
+    return _index_img(imgs, index)
+
+
+def iter_img(imgs):
+    """Iterates over a 4D Niimg-like object in the fourth dimension
+
+    Parameters
+    ----------
+    imgs: 4D Niimg-like object
+          See http://nilearn.github.io/building_blocks/manipulating_mr_images.html#niimg.
+
+    Returns
+    -------
+    output: iterator of 3D nibabel.Nifti1Image
+    """
+    def generator_helper(imgs):
+        imgs = check_niimgs(imgs)
+        # Using a 'yield' so that we can advance until here to raise an
+        # exception if the image is not 4D
+        yield
+
+        for i in range(_get_shape(imgs)[3]):
+            yield _index_img(imgs, i)
+
+    generator = generator_helper(imgs)
+    # This 'next' advances to the first 'yield' and raises an
+    # exception if the image is not 4D. The point of this is to raise
+    # an exception when iter_img is called rather than when the
+    # generator returned by iter_img is used
+    next(generator)
+    return generator
```

Comment:
> class Iterator(object):
>     def __init__(self, imgs):
>         self.imgs = check_niimgs(imgs)
>     def __iter__(self):
>         for i in range(_get_shape(imgs)[3]):
>             yield _index_img(imgs, i)

I actually feel better about this. You will seldom hear me argue for a
class instead of functions, but I feel that the class is more explicite
here.